improve annotation device parameters where a device ordinal is allowed

nairbv opened the pull request:
Using mypy in code that depends on pytorch, I noticed that the type annotation doesn't allow a device ordinal.

`error: Argument "device" to "to_empty" of "Module" has incompatible type "int"; expected "str | device"  [arg-type]`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113647

nairbv committed:
improve annotation - a device ordinal is often used here

[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

albanD commented:
@pytorchbot merge

albanD commented:
Ho looks like types are not matching down either.
I think `int` as a device is not something we ever encourage. So it's possible that types don't have that.

nairbv commented:
I think I just need to update `empty_like` to match `to_empty`?

I can understand preferring not to support `int`, but it works and does seem to be a common usage pattern in code I see out there. FSDP even takes a "device_id" not a device or str.

Skylion007 commented:
Specifically, you may want to see if you can use the DeviceLike TypeAlias and expand that to fix it codebase wide.

nairbv commented:
> use the DeviceLike TypeAlias and expand that to fix it codebase wide.

Will that actually work? I'm assuming there are also places where we don't allow passing an int as a device.

nairbv committed:
Update on "improve annotation torch to_empty / empty_like -- a device ordinal is often used here"

Using mypy in code that depends on pytorch, I noticed that the type annotation doesn't allow a device ordinal.

`error: Argument "device" to "to_empty" of "Module" has incompatible type "int"; expected "str | device"  [arg-type]`




[ghstack-poisoned]

nairbv committed:
Update on "improve annotation torch to_empty / empty_like -- a device ordinal is often used here"

Using mypy in code that depends on pytorch, I noticed that the type annotation doesn't allow a device ordinal.

`error: Argument "device" to "to_empty" of "Module" has incompatible type "int"; expected "str | device"  [arg-type]`




[ghstack-poisoned]

nairbv committed:
Update on "improve annotation torch to_empty / empty_like -- a device ordinal is often used here"

Using mypy in code that depends on pytorch, I noticed that the type annotation doesn't allow a device ordinal.

`error: Argument "device" to "to_empty" of "Module" has incompatible type "int"; expected "str | device"  [arg-type]`




[ghstack-poisoned]

nairbv committed:
Update on "improve annotation torch to_empty / empty_like -- a device ordinal is often used here"

Using mypy in code that depends on pytorch, I noticed that the type annotation doesn't allow a device ordinal.

`error: Argument "device" to "to_empty" of "Module" has incompatible type "int"; expected "str | device"  [arg-type]`




[ghstack-poisoned]

nairbv committed:
Update on "improve annotation torch to_empty / empty_like -- a device ordinal is often used here"

Using mypy in code that depends on pytorch, I noticed that the type annotation doesn't allow a device ordinal.

`error: Argument "device" to "to_empty" of "Module" has incompatible type "int"; expected "str | device"  [arg-type]`




[ghstack-poisoned]

nairbv committed:
Update on "improve annotation device parameters where a device ordinal is allowed"

Using mypy in code that depends on pytorch, I noticed that the type annotation doesn't allow a device ordinal.

`error: Argument "device" to "to_empty" of "Module" has incompatible type "int"; expected "str | device"  [arg-type]`




[ghstack-poisoned]

nairbv committed:
Update on "improve annotation device parameters where a device ordinal is allowed"

Using mypy in code that depends on pytorch, I noticed that the type annotation doesn't allow a device ordinal.

`error: Argument "device" to "to_empty" of "Module" has incompatible type "int"; expected "str | device"  [arg-type]`




[ghstack-poisoned]

nairbv commented:
@albanD I think this is working now, only a few tests still running. It turned out to be a larger change than initially anticipated, so not sure if you want to take another look before merging.

albanD commented on torch/types.py:
```diff
@@ -34,7 +34,7 @@
 # literal device object).  This nomenclature is consistent with PythonArgParser.
 # None means use the default device (typically CPU)
-Device = Union[_device, str, _int, None]
+Device = Optional[Union[_device, str, _int]]
```
Optional[DeviceTypeLike] ? Or that would be a circular import?

nairbv replied:
I didn't actually try it but yeah being a bit cautious I guess since this separate definition already existed. Just adding `Optional` in the import required adding to `test/allowlist_for_publicAPI.json`.

albanD replied:
Nothing should be added to that list, either add a proper `__all__` to this file or `del` this one at the end.

nairbv replied:
del'd the Optional and removed from the allowlist. I also tested with Optional[DeviceLikeType] here in types.py and confirmed it does lead to circular import errors.

albanD reviewed (commented):
Sounds good!
Thanks for taking the time to clean this up!

nairbv committed:
Update on "improve annotation device parameters where a device ordinal is allowed"

Using mypy in code that depends on pytorch, I noticed that the type annotation doesn't allow a device ordinal.

`error: Argument "device" to "to_empty" of "Module" has incompatible type "int"; expected "str | device"  [arg-type]`




[ghstack-poisoned]

albanD reviewed (approved):
Perfect!
Thanks a lot!

albanD commented:
@pytorchbot merge
